usr/man/man/usr/share/man/man1/isainfo.1
'\" te
.\"  Copyright (c) 2007, Sun Microsystems, Inc.  All Rights Reserved
.TH isainfo 1 "8 Feb 2007" "SunOS 5.11" "User Commands"
.SH NAME
isainfo \- describe instruction set architectures
.SH SYNOPSIS
.LP
.nf
\fBisainfo\fR [ [\fB-v\fR] [\fB-b\fR | \fB-n\fR | \fB-k\fR] | [\fB-x\fR]]
.fi

.SH DESCRIPTION
.sp
.LP
The \fBisainfo\fR utility is used to identify various attributes of the instruction set architectures supported on the currently running system. Among the questions it can answer are whether 64-bit applications are supported, or whether the running kernel uses 32-bit or 64-bit device drivers.
.sp
.LP
When invoked with no options, \fBisainfo\fR prints the names of the native instruction sets for applications supported by the current version of the operating system. These are a subset of the list returned by \fBisalist\fR(1). The subset corresponds to the basic applications environments supported by the currently running system.
.SH OPTIONS
.sp
.LP
The following options are supported:
.sp
.ne 2
.mk
.na
\fB\fB-b\fR\fR
.ad
.RS 6n
.rt  
Prints the number of bits in the address space of the native instruction set.
.RE

.sp
.ne 2
.mk
.na
\fB\fB-k\fR\fR
.ad
.RS 6n
.rt  
Prints the name of the instruction set(s) used by the operating system kernel components such as device drivers and \fBSTREAMS\fR modules.
.RE

.sp
.ne 2
.mk
.na
\fB\fB-n\fR\fR
.ad
.RS 6n
.rt  
Prints the name of the native instruction set used by portable applications supported by the current version of the operating system.
.RE

.sp
.ne 2
.mk
.na
\fB\fB-v\fR\fR
.ad
.RS 6n
.rt  
When used with the \fB-b\fR, \fB-k\fR or \fB-n\fR options, prints more detailed information.
.RE

.sp
.ne 2
.mk
.na
\fB\fB-x\fR\fR
.ad
.RS 6n
.rt  
Prints instruction extensions to the native ABI which are supported by the platform.
.RE

.SH EXAMPLES
.LP
\fBExample 1 \fRInvoking \fBisainfo\fR on a 32-bit x86 Platform
.sp
.LP
The following example invokes \fBisainfo\fR on a 32-bit x86 platform:

.sp
.in +2
.nf
example% isainfo -v
32-bit i386 applications

example% isainfo -k
i386
.fi
.in -2
.sp

.LP
\fBExample 2 \fRInvoking \fBisainfo\fR on a System Running the 64-bit Operating System on a 64-bit SPARC Processor
.sp
.LP
The following example invokes \fBisainfo\fR on a system running the 64-bit operating system on a 64-bit SPARC processor:

.sp
.in +2
.nf
example% isainfo
sparcv9 sparc
example% isainfo -n
sparcv9
example% isainfo -v
64-bit sparcv9 applications
32-bit sparc applications
example% isainfo -vk
64-bit sparcv9 kernel modules
.fi
.in -2
.sp

.LP
\fBExample 3 \fRInvoking \fBisainfo\fR \fB-x\fR on an AMD Opteron CPU
.sp
.LP
The following example invokes \fBisainfo\fR with the \fB-x\fR option on an AMD Opteron CPU: 

.sp
.in +2
.nf
example% isainfo -x
i386: fpu tsc cx8 sep cmov mmx ammx a3dnow a3dnowx fxsr sse sse2 pause
.fi
.in -2
.sp

.SH EXIT STATUS
.sp
.ne 2
.mk
.na
\fBNon-zero\fR
.ad
.RS 12n
.rt  
Options are not specified correctly, or the command is unable to recognize attributes of the system on which it is running. An error message is printed to \fBstderr\fR.
.RE

.SH ATTRIBUTES
.sp
.LP
See \fBattributes\fR(5) for descriptions of the following attributes:
.sp

.sp
.TS
tab() box;
cw(2.75i) |cw(2.75i) 
lw(2.75i) |lw(2.75i) 
.
ATTRIBUTE TYPEATTRIBUTE VALUE
_
Availabilitysystem/core-os
.TE

.SH SEE ALSO
.sp
.LP
\fBisalist\fR(1), \fBuname\fR(1), \fBpsrinfo\fR(1M), \fBgetisax\fR(2), \fBsysinfo\fR(2), \fBattributes\fR(5), \fBisalist\fR(5)


usr/src/cmd/isainfo/isainfo.c
/*
 * CDDL HEADER START
 *
 * The contents of this file are subject to the terms of the
 * Common Development and Distribution License (the "License").
 * You may not use this file except in compliance with the License.
 *
 * You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
 * or http://www.opensolaris.org/os/licensing.
 * See the License for the specific language governing permissions
 * and limitations under the License.
 *
 * When distributing Covered Code, include this CDDL HEADER in each
 * file and include the License file at usr/src/OPENSOLARIS.LICENSE.
 * If applicable, add the following below this CDDL HEADER, with the
 * fields enclosed by brackets "[]" replaced with your own identifying
 * information: Portions Copyright [yyyy] [name of copyright owner]
 *
 * CDDL HEADER END
 */
/*
 * Copyright (c) 1998, 2011, Oracle and/or its affiliates. All rights reserved.
 */

#include <sys/types.h>
#include <sys/systeminfo.h>
#include <sys/utsname.h>
#include <sys/stat.h>

#include <sys/auxv.h>
#include <sys/cpuid_drv.h>
#include <sys/elf.h>

#include <stdio.h>
#include <stdlib.h>
#include <strings.h>
#include <unistd.h>
#include <errno.h>
#include <libintl.h>
#include <locale.h>
#include <fcntl.h>

#include <elfcap.h>

static const char dev_cpu_self_cpuid[] = "/dev/" CPUID_SELF_NAME;
static char *pgmname;
static int mode = 0;

#define	BITS_MODE	0x1
#define	NATIVE_MODE	0x2
#define	KERN_MODE	0x4
#define	VERBOSE_MODE	0x8
#define	EXTN_MODE	0x10

static char *
getsysinfo(int cmd)
{
	char *buf;
	size_t bufsize = 20;	/* wild guess */
	long ret;

	if ((buf = malloc(bufsize)) == NULL) {
		(void) fprintf(stderr,
		    gettext("%s: malloc failed \n"), pgmname);
		return (NULL);
	}
	do {
		ret = sysinfo(cmd, buf, bufsize);
		if (ret == -1) {
			(void) fprintf(stderr,
			    gettext("%s: sysinfo failed\n"), pgmname);
			return (NULL);
		}
		if (ret > bufsize) {
			bufsize = ret;
			if ((buf = realloc(buf, bufsize)) == NULL) {
				(void) fprintf(stderr,
				gettext("%s: realloc failed \n"), pgmname);
				return (NULL);
			}
		} else
			break;
	} while (buf != NULL);

	return (buf);
}

/*
 * Classify isa's as to bitness of the corresponding ABIs.
 * isa's which have no "official" Solaris ABI are returned
 * unrecognised i.e. "zero bit".
 */
static uint_t
bitness(const char *isaname)
{
	if (strcmp(isaname, "sparc") == 0 ||
	    strcmp(isaname, "i386") == 0)
		return (32);

	if (strcmp(isaname, "sparcv9") == 0 ||
	    strcmp(isaname, "amd64") == 0)
		return (64);

	return (0);
}

static char *
report_abi(int cmd, const char *vfmt)
{
	uint_t bits;
	char *isa;

	if ((isa = getsysinfo(cmd)) == NULL)
		exit(2);
	if ((bits = bitness(isa)) == 0) {
		(void) fprintf(stderr,
		    gettext("%s: unable to identify isa '%s'!\n"),
		    pgmname, isa);
		exit(3);
	}

	if (mode & VERBOSE_MODE)
		(void) printf(vfmt, bits, isa);
	else if (mode & BITS_MODE)
		(void) printf("%d\n", bits);
	else if (mode & (NATIVE_MODE|KERN_MODE))
		(void) printf("%s\n", isa);
	else
		(void) printf("%s", isa);
	return (isa);
}

/*
 * Classify isas as their machine type.
 */
static ushort_t
machtype(const char *isaname)
{
	if (strcmp(isaname, "sparc") == 0)
		return (EM_SPARC);
	if (strcmp(isaname, "sparcv9") == 0)
		return (EM_SPARCV9);
	if (strcmp(isaname, "i386") == 0)
		return (EM_386);
	if (strcmp(isaname, "amd64") == 0)
		return (EM_AMD64);

	return (0);
}

static void
report_hwcap(int d, const char *isa)
{
	struct cpuid_get_hwcap __cgh, *cgh = &__cgh;
	char buffer[1024];

	cgh->cgh_archname = (char *)isa;
	if (ioctl(d, CPUID_GET_HWCAP, cgh) != 0)
		return;

	(void) elfcap_hw1_to_str(ELFCAP_STYLE_LC, cgh->cgh_hwcap,
	    buffer, sizeof (buffer), ELFCAP_FMT_SNGSPACE, machtype(isa));

	if (mode & EXTN_MODE) {
		(void) printf(": %s\n", buffer);
	} else {
		char *p;
		int linecnt = 0;

		for (p = strtok(buffer, " "); p; p = strtok(NULL, " ")) {
			if (linecnt == 0)
				linecnt = printf("\t");
			linecnt += printf("%s ", p);
			if (linecnt > 68) {
				(void) printf("\n");
				linecnt = 0;
			}
		}
		if (linecnt != 0)
			(void) printf("\n");
	}
}

#if !defined(TEXT_DOMAIN)
#define	TEXT_DOMAIN	"SYS_TEST"
#endif

int
main(int argc, char *argv[])
{
	int errflg = 0;
	int c;
	char *vfmt;
	char *isa, *isa32;
	int d = -1;
	const int excl_modes =	/* exclusive mode settings */
	    NATIVE_MODE | BITS_MODE | KERN_MODE | EXTN_MODE;

	(void) setlocale(LC_ALL, "");
	(void) textdomain(TEXT_DOMAIN);

	if ((pgmname = strrchr(*argv, '/')) == 0)
		pgmname = argv[0];
	else
		pgmname++;

	while ((c = getopt(argc, argv, "nbkvx")) != EOF)
		switch (c) {
		case 'n':
			if (mode & excl_modes)
				errflg++;
			mode |= NATIVE_MODE;
			break;
		case 'b':
			if (mode & excl_modes)
				errflg++;
			mode |= BITS_MODE;
			break;
		case 'k':
			if (mode & excl_modes)
				errflg++;
			mode |= KERN_MODE;
			break;
		case 'x':
			if (mode & excl_modes || mode & VERBOSE_MODE)
				errflg++;
			mode |= EXTN_MODE;
			break;
		case 'v':
			if (mode & EXTN_MODE)
				errflg++;
			mode |= VERBOSE_MODE;
			break;
		case '?':
		default:
			errflg++;
			break;
		}

	if (errflg || optind != argc) {
		(void) fprintf(stderr,
		    gettext("usage: %s [ [-v] [-b | -n | -k] | [-x] ]\n"),
		    pgmname);
		return (1);
	}

	/*
	 * We use dev_cpu_self_cpuid for discovering hardware capabilities;
	 * but we only complain if we can't open it if we've been
	 * asked to report on those capabilities.
	 */
	if ((mode & (VERBOSE_MODE|EXTN_MODE)) != 0 &&
	    (d = open(dev_cpu_self_cpuid, O_RDONLY)) == -1)
		perror(dev_cpu_self_cpuid), exit(1);

	if (mode & KERN_MODE) {
		vfmt = gettext("%d-bit %s kernel modules\n");
		(void) report_abi(SI_ARCHITECTURE_K, vfmt);
		return (0);
	}

	vfmt = gettext("%d-bit %s applications\n");

	if (mode & (BITS_MODE | NATIVE_MODE)) {
		if ((isa = report_abi(SI_ARCHITECTURE_64, vfmt)) == NULL)
			isa = report_abi(SI_ARCHITECTURE_32, vfmt);
		if (isa != NULL && (mode & VERBOSE_MODE) != 0)
			report_hwcap(d, isa);
	} else {
		if ((isa = report_abi(SI_ARCHITECTURE_64, vfmt)) != NULL) {
			if (mode & (EXTN_MODE|VERBOSE_MODE))
				report_hwcap(d, isa);
			else
				(void) putchar(' ');
		}

		if ((isa32 = report_abi(SI_ARCHITECTURE_32, vfmt)) != NULL) {
			if (mode & (EXTN_MODE|VERBOSE_MODE))
				report_hwcap(d, isa32);
		}

		if ((isa32 != NULL || isa != NULL) &&
		    (mode & (EXTN_MODE|VERBOSE_MODE)) == 0)
			(void) putchar('\n');
	}

	return (0);
}
